src/NHibernate/Impl/SessionImpl.cs

```diff
@@ -1565,29 +1566,27 @@ public void Dispose()
 			using (new SessionIdLoggingContext(SessionId))
 			{
 				log.Debug(string.Format("[session-id={0}] running ISession.Dispose()", SessionId));
+				TransactionContext?.Wait();
 				var transactionContext = TransactionContext;
-				if (transactionContext != null)
+				if (transactionContext != null && transactionContext.CanFlushOnSystemTransactionCompleted)
 				{
-					transactionContext.Wait();
-					if (transactionContext.CanFlushOnSystemTransactionCompleted)
+					// absolute nogo to remove before merge, done for illustrating a race condition
+					try
 					{
-						// absolute nogo to remove before merge, done for illustrating a race condition
-						try
-						{
-							// Bloat only distributed case
-							if (((Transaction.AdoNetWithSystemTransactionFactory.SystemTransactionContext)transactionContext)
-							    .AmbientTransaction.TransactionInformation.DistributedIdentifier != Guid.Empty)
-								System.Threading.Thread.Sleep(500);
-						}
-						catch
-						{
-							// do not care
-						}
-						// end absolute nogo
-						transactionContext.ShouldCloseSessionOnSystemTransactionCompleted = true;
-						return;
+						// Bloat only distributed case
+						if (((Transaction.AdoNetWithSystemTransactionFactory.SystemTransactionContext) transactionContext)
```

Comment:
Still not good, must take again the property. Better revert your changes and go back to the original code. Maybe it was lacking some comments, but it was written that way for good reasons.